File: tools/gltf_to_glb.js
const fs = require('fs');
const path = require('path');

const root = path.resolve(__dirname, '..');
const modelDir = path.join(root, 'handshake_-_ramadhan_series');
const inputGltf = path.join(modelDir, 'scene.gltf');

if (!fs.existsSync(inputGltf)) {
  console.error('scene.gltf not found at', inputGltf);
  process.exit(1);
}

const gltf = JSON.parse(fs.readFileSync(inputGltf, 'utf8'));
const buffers = gltf.buffers || [];
if (!buffers.length) { console.error('No buffers'); process.exit(1); }
const bufferUri = buffers[0].uri;
if (!bufferUri) { console.error('No buffer uri'); process.exit(1); }
const binPath = path.join(modelDir, bufferUri);
if (!fs.existsSync(binPath)) { console.error('Binary buffer not found:', binPath); process.exit(1); }
const binBytes = fs.readFileSync(binPath);

// remove uri
delete buffers[0].uri;

let jsonText = JSON.stringify(gltf);
let jsonBytes = Buffer.from(jsonText, 'utf8');
const jsonPad = (4 - (jsonBytes.length % 4)) % 4;
if (jsonPad) jsonBytes = Buffer.concat([jsonBytes, Buffer.alloc(jsonPad, 0x20)]);

const binPad = (4 - (binBytes.length % 4)) % 4;
let binBytesPadded = binBytes;
if (binPad) binBytesPadded = Buffer.concat([binBytes, Buffer.alloc(binPad, 0x00)]);

const jsonChunkLength = jsonBytes.length;
const binChunkLength = binBytesPadded.length;
const totalLength = 12 + 8 + jsonChunkLength + 8 + binChunkLength;

const header = Buffer.alloc(12);
header.write('glTF', 0, 4, 'ascii');
header.writeUInt32LE(2, 4);
header.writeUInt32LE(totalLength, 8);

const jsonHeader = Buffer.alloc(8);
jsonHeader.writeUInt32LE(jsonChunkLength, 0);
jsonHeader.write('JSON', 4, 4, 'ascii');

const binHeader = Buffer.alloc(8);
binHeader.writeUInt32LE(binChunkLength, 0);
binHeader.write('BIN\x00', 4, 4, 'ascii');

const outPath = path.join(modelDir, 'scene.glb');
const out = Buffer.concat([header, jsonHeader, jsonBytes, binHeader, binBytesPadded]);
fs.writeFileSync(outPath, out);
console.log('Wrote', outPath, out.length);

File: tools/gltf_to_glb.py
#!/usr/bin/env python3
import json
import os
import struct

root = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
model_dir = os.path.join(root, 'handshake_-_ramadhan_series')
input_gltf = os.path.join(model_dir, 'scene.gltf')

with open(input_gltf, 'r', encoding='utf-8') as f:
    gltf = json.load(f)

buffers = gltf.get('buffers', [])
if not buffers:
    raise SystemExit('No buffers found in GLTF')

buffer_uri = buffers[0].get('uri')
if not buffer_uri:
    raise SystemExit('Buffer has no uri; glTF may already be binary')

bin_path = os.path.join(model_dir, buffer_uri)
if not os.path.exists(bin_path):
    raise SystemExit(f'Binary buffer not found: {bin_path}')

with open(bin_path, 'rb') as bf:
    bin_bytes = bf.read()

# Remove uri from buffers so GLB points to binary chunk
buffers[0].pop('uri', None)

# Encode JSON chunk
json_text = json.dumps(gltf, separators=(',', ':'), ensure_ascii=False)
json_bytes = json_text.encode('utf-8')
# pad to 4 bytes
json_padding = (4 - (len(json_bytes) % 4)) % 4
json_bytes += b' ' * json_padding

# pad binary to 4 bytes
bin_padding = (4 - (len(bin_bytes) % 4)) % 4
bin_bytes_padded = bin_bytes + (b"\x00" * bin_padding)

json_chunk_len = len(json_bytes)
bin_chunk_len = len(bin_bytes_padded)

# GLB header: magic, version(2), length (placeholder)
# total length = 12 (header) + 8 + json_chunk_len + 8 + bin_chunk_len
total_length = 12 + 8 + json_chunk_len + 8 + bin_chunk_len

glb = bytearray()
glb += b'glTF'                    # magic
glb += struct.pack('<I', 2)       # version
glb += struct.pack('<I', total_length)

# JSON chunk header
glb += struct.pack('<I', json_chunk_len)
glb += b'JSON'
glb += json_bytes

# BIN chunk header
glb += struct.pack('<I', bin_chunk_len)
glb += b'BIN\x00'
glb += bin_bytes_padded

out_path = os.path.join(model_dir, 'scene.glb')
with open(out_path, 'wb') as out:
    out.write(glb)

print('Wrote', out_path, 'size', os.path.getsize(out_path))
